GitUI/BranchTreePanel/BaseRefTree.cs

```diff
@@ -80,5 +84,56 @@ internal override void UpdateVisibility()
 
             base.UpdateVisibility();
         }
+
+        protected IEnumerable<IGitRef> PriotizedBranches(IReadOnlyList<IGitRef> branches)
+        {
+            return OrderByPriority(branches, node => node.LocalName, AppSettings.PrioritizedBranchNames);
+        }
+
+        protected IEnumerable<RemoteRepoNode> PriotizedRemotes(IReadOnlyList<RemoteRepoNode> remotes)
+        {
+            return OrderByPriority(remotes.OrderBy(node => node.FullPath), node => node.FullPath, AppSettings.PrioritizedRemoteNames);
+        }
+
+        protected IEnumerable<Remote> PriotizedRemotes(IReadOnlyList<Remote> remotes)
+        {
+            return OrderByPriority(remotes.OrderBy(node => node.Name), node => node.Name, AppSettings.PrioritizedRemoteNames);
+        }
+
+        /// <summary>
+        /// Order the references by the prioties in the setting regex
+        /// </summary>
+        /// <typeparam name="T">The type to prioritize, e.g. IGitRef.</typeparam>
+        /// <param name="references">The branches or remotes to prioritize.</param>
+        /// <param name="keySelector">Function in T to get the sorter.</param>
+        /// <param name="setting">String with regexes with priorities separated by semicolon.</param>
+        /// <returns>The resorted references.</returns>
+        private IEnumerable<T> OrderByPriority<T>(IEnumerable<T> references, Func<T, string> keySelector, string setting)
+        {
+            // Sort prio branches first (if set) with the compile cache (no need to instanceiate)
+            var regexes = setting.Split(";", StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
+            if (!regexes.Any())
+            {
+                return references;
+            }
+
+            int currPrio = 0;
+            Dictionary<string, int> prios = new();
+            foreach (string regex in regexes.Reverse())
+            {
+                currPrio++;
+                references.ForEach(node =>
+                {
+                    string key = keySelector(node);
+                    if (Regex.IsMatch(key, $"^({regex})$", RegexOptions.ExplicitCapture))
+                    {
+                        prios[key] = currPrio;
+                    }
+                });
+            }
+
+            // Set prio on all branches, let no match be 0
+            return references.OrderBy(node => prios.TryGetValue(keySelector(node), out int prio) ? -prio : 0);
```

Comment:
We're re-enumerating the underlying collection; this is not good. `references` has to be a materialised collection like `List<>`.

Reply:
The list is materialized.
The core performance is required for the regex second for the sort anyway.

Reply:
With 500 local branches and 3 prio levels, this add a few ms for local branches first times, no noticable time for refresh while the 15 entry static cache in .NET is utilized
https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices

Reply:
> The list is materialized.

No. `remotes.OrderBy(node => node.FullPath)` is an enumeration which isn't materialised. We enumerate the underlying collection once at line:120 and then re-enumerate it again at line:131.

Reply:
references is materialised by the caller (it is in the "cache").

For the output: Do you want to return List<T> with ToList() in this function or let the users call ToList()?

Reply:
> references is materialised by the caller (it is in the "cache").

I can't see that happening.
```cs
        private IEnumerable<T> OrderByPriority<T>(IEnumerable<T> references...)
        {
            // ...

            if (!regexes.Any())
            {
// returns the non-enumerated dataset
                return references;
            }

            // ...

// we're now enumerating the underlying dataset
            foreach (T node in references)
            {
                // ...
            }

// and now we're enumerating the underlying dataset again!
            // Order by the sort match, with no match last as int.Max
            return references.OrderBy(node => orderByNodeKey.GetValueOrDefault(keySelector(node), int.MaxValue));
        }
```